### PR TITLE
Unittest: add a test folder v1

### PR DESCRIPTION
**Pull request by audoan99**
This folder consists of 3 parts:
- client: test ivi-layermanagement-api (ilmCommon, ilmControl, ilmInput).
- server: test ivi-id-agent-modules, ivi-input-modules and weston-ivi-shell.
- common: contains common fake APIs for both client and server.

Struct of tests:
image

### CODE REVIEW

File: unittest/client/src/ilm_common_unittests.cpp [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    SET_RETURN_SEQ(ilmControl_init, mp_ilmErrorType, MAX_NUMBER);
+    // Invoke the ilmControl_init and expect the result is ILM_FAILED
+    ASSERT_EQ(ILM_FAILED, ilm_init());
+    /* Check the logic
+     * ilmControl_init_fake should call one time and return ILM_FAILED
+     */
+    ASSERT_EQ(ilmControl_init_fake.call_count, 1);
+    ASSERT_EQ(mp_ilmErrorType[0], ilmControl_init_fake.return_val_history[0]);
+}
+
```
Comment (HarshaMM) [inline unittest/client/src/ilm_common_unittests.cpp:96]: It would be beneficial to have a concise description and goal of each test case.

File: unittest/client/src/ilm_common_unittests.cpp [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    ASSERT_EQ(wl_display_connect_fake.call_count, 1);
+    ASSERT_EQ(ilmControl_init_fake.call_count, 1);
+    ASSERT_EQ(ILM_SUCCESS, ilmControl_init_fake.return_val_history[0]);
+}
+
+TEST_F(IlmCommonTest, ilm_init_manyTimes)
+{
+    // Prepare fake for wl_display_connect and ilmControl_init, to make the ilm_init return success
+    mock_ilmInitSuccess();
+    // Invoke the ilm_init in first time and expect return ILM_SUCCESS
```
Comment (HarshaMM) [inline unittest/client/src/ilm_common_unittests.cpp:116]: I see too many comments in the code. Its really not necessary. e.g. Individual comments in this function can be concisely written as test for multiple ilm_init and ilm_destroy calls.

File: unittest/client/src/ilm_common_unittests.cpp (hunk 1 above)
Comment (audoan99) [reply unittest/client/src/ilm_common_unittests.cpp:106]: Hi @HarshaMM, Thanks for your comment.
I added a concise description and goal to some test cases and remove comments inside tests. Please review it for me.
And I will change it to all of the tests if you approve.

File: unittest/client/src/ilm_common_unittests.cpp [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+    static constexpr uint8_t MAX_NUMBER = 1;
+    uint8_t m_wlDisplayFakePointer = 0;
+    struct wl_display* mpp_wlDisplays [MAX_NUMBER] = {nullptr};
+    ilmErrorTypes mp_ilmErrorType[MAX_NUMBER] = {ILM_FAILED};
+};
+
+/** ================================================================================================
+ * @test_id             ilm_init_cannotGetDisplay
+ * @brief               A failure of ilm_init() when wl_display_connect() returns a null object
```
Comment (efriedrich) [inline unittest/client/src/ilm_common_unittests.cpp:68]: I would rephrase this to: Test case of ilm_init() where wl_display_connect fails, returns null object

File: unittest/client/src/ilm_common_unittests.cpp [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+    mpp_wlDisplays[0] = nullptr;
+    SET_RETURN_SEQ(wl_display_connect, mpp_wlDisplays, MAX_NUMBER);
+    ASSERT_EQ(ILM_FAILED, ilm_init());
+    ASSERT_EQ(wl_display_connect_fake.call_count, 1);
+    ASSERT_EQ(wl_display_connect_fake.return_val_history[0], nullptr);
+}
+
+/** ================================================================================================
+ * @test_id             ilm_init_getFailedOnilmControl_init
+ * @brief               A failure of ilm_init() when ilmControl_init() returns ILM_FAILED
```
Comment (efriedrich) [inline unittest/client/src/ilm_common_unittests.cpp:87]: this I would also rephrase with the similar pattern: Test case of ilm_init() where the internal function ilmControl_init() fails

File: unittest/client/src/ilm_common_unittests.cpp [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    mp_ilmErrorType[0] = ILM_FAILED;
+    SET_RETURN_SEQ(ilmControl_init, mp_ilmErrorType, MAX_NUMBER);
+    ASSERT_EQ(ILM_FAILED, ilm_init());
+    ASSERT_EQ(ilmControl_init_fake.call_count, 1);
+    ASSERT_EQ(mp_ilmErrorType[0], ilmControl_init_fake.return_val_history[0]);
+}
+
+/** ================================================================================================
+ * @test_id             ilm_init_getSuccess
+ * @brief               A success of ilm_init() when wl_display_connect() doesn't return null object
```
Comment (efriedrich) [inline unittest/client/src/ilm_common_unittests.cpp:109]: please follow the same wording pattern as above

File: unittest/client/src/ilm_common_unittests.cpp [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    mock_ilmInitSuccess();
+    ASSERT_EQ(ILM_SUCCESS, ilm_init());
+    ASSERT_EQ(wl_display_connect_fake.call_count, 1);
+    ASSERT_EQ(ilmControl_init_fake.call_count, 1);
+    ASSERT_EQ(ILM_SUCCESS, ilmControl_init_fake.return_val_history[0]);
+}
+
+/** ================================================================================================
+ * @test_id             ilm_init_manyTimes
+ * @brief               The number of calls to ilm_init() and ilm_destroy() must correspond.
```
Comment (efriedrich) [inline unittest/client/src/ilm_common_unittests.cpp:131]: rephrase to something like: Test case to ensure the number of calls to ilm_init() ...

**Comment by audoan99** [comment]
Tested-by: Au Doan Ngoc [au.doanngoc@vn.bosch.com](mailto:au.doanngoc@vn.bosch.com)

File: unittest/server/src/ivi_controller_uinttests.cpp [hunk 7]
```diff
@@ -0,0 +1,3361 @@
...
+ *                      -# Mocking the surface_dump() does return -1
+ *                      -# Calling the controller_surface_screenshot()
+ *                      -# Verification point:
+ *                         +# wl_resource_post_event() must be called once time
+ *                         +# wl_resource_destroy() must be called once time
+ */
+TEST_F(ControllerTests, controller_surface_screenshot_errReadpixByResult)
+{
+    SET_RETURN_SEQ(wl_resource_get_user_data, (void**)&mp_iviController[0], 1);
+    struct wl_resource * screenshot[1];
```
Comment (khangtb1) [inline unittest/server/src/ivi_controller_uinttests.cpp:2358]: As a discussion: the variable only declare without initial, it may make a potential issue when getting random value. Please take a look to all variables without initialize.

File: unittest/server/src/ivi_controller_uinttests.cpp [hunk 8]
```diff
@@ -0,0 +1,3361 @@
...
+    SET_RETURN_SEQ(wl_resource_create, &screenshot[0], 1);
```
Comment (khangtb1) [inline unittest/server/src/ivi_controller_uinttests.cpp:2359]: The second argument of SET_RETURN_SEQ is the ARRAY of the return value type of fake function. this is still correct, but should keep it more simple: `SET_RETURN_SEQ(wl_resource_create, screenshot, 1);`

File: unittest/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+############################################################################
+
+CMAKE_MINIMUM_REQUIRED(VERSION 2.6)
+
```
Comment (khangtb1) [inline unittest/CMakeLists.txt:21]: The warning messages when configuring of cmake.
change to: https://github.com/COVESA/wayland-ivi-extension/pull/165/commits/65fa39b1b1edf8806f9bbfacaea07fae12347b26

File: unittest/client/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+############################################################################
+
+CMAKE_MINIMUM_REQUIRED(VERSION 2.6)
+cmake_policy(SET CMP0046 OLD)
+
```
Comment (khangtb1) [inline unittest/client/CMakeLists.txt:22]: The warning messages when configuring of cmake.
https://github.com/COVESA/wayland-ivi-extension/pull/165/commits/65fa39b1b1edf8806f9bbfacaea07fae12347b26

File: unittest/client/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    ${CMAKE_CURRENT_BINARY_DIR}/../../protocol
+    ${CMAKE_CURRENT_SOURCE_DIR}/include
+    ${CMAKE_CURRENT_SOURCE_DIR}/../common
+    ${WAYLAND_CLIENT_INCLUDE_DIRS}
+    ${WESTON_INCLUDE_DIRS}
+    ${gtest_INCLUDE_DIRS}
+)
+
+LINK_DIRECTORIES(
+    ${WAYLAND_CLIENT_LIBRARY_DIRS}
```
Comment (khangtb1) [inline unittest/client/CMakeLists.txt:40]: no binaries link to libwayland-client.so, don't need to keep this link

File: unittest/client/src/ilm_control_unittests.cpp [hunk 12]
```diff
@@ -0,0 +1,5071 @@
...
+        g_ilmControlStatus = NONE;
+        g_ilm_notification_mask = ILM_NOTIFICATION_ALL;
+    }
+
+    void TearDown()
+    {
+        deinit_ctx_list_content();
+    }
+
+    void init_ctx_list_content()
```
Comment (khangtb1) [inline unittest/client/src/ilm_control_unittests.cpp:151]: consider to divide this function to sub-functions, for supporting add new surface, seat, layer, and screen. it will be cleaner than currently.

File: unittest/client/src/ilm_control_unittests.cpp [hunk 13]
```diff
@@ -0,0 +1,5071 @@
...
+TEST_F(IlmControlTest, ilm_getPropertiesOfSurface_cannotGetRoundTripQueue)
+{
+    SET_RETURN_SEQ(wl_display_roundtrip_queue, mp_failureResult, 1);
+
+    struct ilmSurfaceProperties *l_surfaceProp = (struct ilmSurfaceProperties *)0xFFFFFFFF;
+    ASSERT_EQ(ILM_FAILED, ilm_getPropertiesOfSurface(1, l_surfaceProp));
+
+    ASSERT_EQ(1, wl_proxy_marshal_flags_fake.call_count);
+    ASSERT_EQ(1, wl_display_roundtrip_queue_fake.call_count);
+    ASSERT_EQ(mp_failureResult[0], wl_display_roundtrip_queue_fake.return_val_history[0]);
```
Comment (khangtb1) [inline unittest/client/src/ilm_control_unittests.cpp:352]: Consider to remove the check the result, that were faked with `SET_RETURN_SEQ`

**Comment by chienpv1** [comment]
To make more convenient to review and update, all changes will continue to be updated in [PR_180](https://github.com/COVESA/wayland-ivi-extension/pull/180). This PR will be closed.

**Comment by efriedrich** [comment]
please just ensure the useful content is also part of https://github.com/COVESA/wayland-ivi-extension/pull/180 and we will continue there....
